developer_docs/codegen_stage.md

```diff
@@ -0,0 +1,281 @@
+E.g:
+```python
+def f():
+    def g():
+        pass
+```
+becomes:
+```c
+void f() {
+}
+
+void f__g() {
+}
+```
+
+Care must be taken regarding any variables local to the enclosing function which are used in the nested function.
+These variables are noted as global variables in the nested function, however as they are not entirely global they must be passed as arguments.
+If they are not modified then they can be passed as normal arguments, however if they are modified they must be passed as pointers and be saved in `CCodePrinter._additional_args` (see [Multiple returns](#multiple-returns)).
+
+### Arrays
+
+Unlike Fortran, C does not have any native support for arrays.
+In order to translate array expressions, Pyccel contains a basic array implementation.
+This implementation can be found in the folder [pyccel/stdlib/ndarrays/](..pyccel/stdlib/ndarrays/).
+It is heavily inspired by the NumPy implementation which makes it easy to collect the array from the NumPy object and pass it to the function.
+
+When adding new array functionalities to Pyccel developers must therefore consider whether it is better/more readable to implement the function in the Pyccel's standard library (`stdlib`) or in the generated code.
+Functions added to the standard library can be more easily optimised, however such implementations can easily lead to large amounts of code duplication, especially if the function involves data manipulation and supports arguments such as `axis`.
+On the other hand implementations written directly in the generated code can decrease readability of the code (try translating the print of an array for an example).
+
+In the future we hope to add an additional option to avoid some of these problems where an implementation of common low-level functions (e.g. `sum`) is written in Python and Pyccel is used to provide the language-specific equivalent of this function with the required argument types.
+This option requires some discussion amongst the senior developers and some additional functions (e.g. `reshape`/`ravel`/`flatten`).
+
+## Fortran-specific problems
+
+### 0-based arrays
+
+Unlike in the majority of coding languages, Fortran array indexes do not have to start from 0.
+In fact, the default start index is 1.
+As a result this needs to be taken into consideration when writing code.
+
+There are two possibilities to handle this; either we can modify every index by adding 1, or we can define our arrays in Fortran with syntax which ensures that the indexes start from 0.
+In Pyccel we have chosen the latter option.
+It is therefore very important to think about this when allocating arrays or pointers.
+
+### Ordering
+
+Different languages handle multi-dimensional arrays in different ways.
+This means that optimal indexing is not written in the same way in different languages.
+Specifically C uses row-major indexing, while Fortran uses column-major indexing.
+Python allows both indexing conventions, but it defaults to C ordering.
```

Comment:
I would change "Python" with "NumPy" in this sentence.

Reply:
Fixed 6230f785